feat: add pkce-openid-backend

johanv26 opened the pull request:
<!--
Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://www.conventionalcommits.org/en/v1.0.0/

Use this template as a guide. Omit sections that don't apply.

🙈 Don't be lazy, try to fill out the template well.
-->

## Description
This adds a generic backend based on ConfigurableOpenIdConnectAuth but
with PKCE.
This backend is inspired by the social-core way to implement PKCE.
There is a current PR in work, but for the moment, that class is not merged and accessible.
So after that is finished this has its code for `code_challenge` and `code_challenge_method`implementation.
PR: https://github.com/python-social-auth/social-core/pull/856
## Testing instructions

Please provide detailed step-by-step instructions for testing this change.

## Additional information

**Jira story**
https://edunext.atlassian.net/browse/FUTUREX-606
## Checklist for Merge

- [ ] Tested in a remote environment
- [ ] Updated documentation
- [ ] Rebased master/main
- [ ] Squashed commits

<!--
You can put NA in the checklist if it doesn't apply

- [x] Check that dont't apply / NA
-->

johanv26 commented:
omar-nelc could you take a look 😬

omar-nelc commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,72 @@ def get_user_id(self, *args, **kwargs):
                 LOG.info("Updating uid: %s to %s", uid, slug_uid)
 
         return slug_uid
+
```
```suggestion
# TODO: Use the `from social_core.backends.oauth import BaseOAuth2PKCE` base class once the pull request is merged: https://github.com/python-social-auth/social-core/pull/856/files#diff-d44db201b48f2ec7cab2a0c981213a2991630567778cc6608d03fa0e3804e466R467
class BaseOAuth2PKCEMixin:
    """
    Base class for providers using OAuth2 with Proof Key for Code Exchange (PKCE).
    OAuth2 details at:
        https://datatracker.ietf.org/doc/html/rfc6749
    PKCE details at:
        https://datatracker.ietf.org/doc/html/rfc7636
    """

    PKCE_DEFAULT_CODE_CHALLENGE_METHOD = "s256"
    PKCE_DEFAULT_CODE_VERIFIER_LENGTH = 32
    USE_PKCE = True

    def create_code_verifier(self):
        name = self.name + "_code_verifier"
        code_verifier_len = self.setting(
            "PKCE_CODE_VERIFIER_LENGTH", default=self.PKCE_DEFAULT_CODE_VERIFIER_LENGTH
        )
        code_verifier = self.strategy.random_string(code_verifier_len)
        self.strategy.session_set(name, code_verifier)
        return code_verifier

    def get_code_verifier(self):
        name = self.name + "_code_verifier"
        code_verifier = self.strategy.session_get(name)
        return code_verifier

    def generate_code_challenge(self, code_verifier, challenge_method):
        method = challenge_method.lower()
        if method == "s256":
            hashed = hashlib.sha256(code_verifier.encode()).digest()
            encoded = base64.urlsafe_b64encode(hashed)
            code_challenge = encoded.decode().replace("=", "")  # remove padding
            return code_challenge
        elif method == "plain":
            return code_verifier
        else:
            raise AuthException("Unsupported code challenge method.")

    def auth_params(self, state=None):
        params = super().auth_params(state=state)

        if self.USE_PKCE:
            code_challenge_method = self.setting(
                "PKCE_CODE_CHALLENGE_METHOD",
                default=self.PKCE_DEFAULT_CODE_CHALLENGE_METHOD,
            )
            code_verifier = self.create_code_verifier()
            code_challenge = self.generate_code_challenge(
                code_verifier, code_challenge_method
            )
            params["code_challenge_method"] = code_challenge_method
            params["code_challenge"] = code_challenge
        return params

    def auth_complete_params(self, state=None):
        params = super().auth_complete_params(state=state)

        if self.USE_PKCE:
            code_verifier = self.get_code_verifier()
            params["code_verifier"] = code_verifier

        return params
```

omar-nelc commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,72 @@ def get_user_id(self, *args, **kwargs):
         return slug_uid
+
+
+class ConfigurableOpenIdConnectAuthPKCE(ConfigurableOpenIdConnectAuth):
```
```suggestion
class ConfigurableOpenIdConnectAuthPKCE(BaseOAuth2PKCEMixin, ConfigurableOpenIdConnectAuth):
```

omar-nelc reviewed (approved):
Thanks @johanv26. I have added a suggestion to refactor the `BaseOAuth2PKCEMixin` and use it so we can refactor later and use the `BaseOAuth2PKCE` base class once it lands on upstream.

This usually helps with the release upgrade.

omar-nelc commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,86 @@ def get_user_id(self, *args, **kwargs):
+
+    """
+    name = 'config-based-openidconnect-PKCE'
+    PKCE_DEFAULT_CODE_CHALLENGE_METHOD = "s256"
+    PKCE_DEFAULT_CODE_VERIFIER_LENGTH = 32
+    USE_PKCE = True
```
Thanks @johanv26!! Looks much better now!

Those aren't needed, because they're inherited from the mixin.

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+    USE_PKCE = True
+
+    def create_code_verifier(self):
+        name = self.name + "_code_verifier"
```
```suggestion
        name = f"{self.name}_code_verifier"
```

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+            "PKCE_CODE_VERIFIER_LENGTH", default=self.PKCE_DEFAULT_CODE_VERIFIER_LENGTH
+        )
+        code_verifier = self.strategy.random_string(code_verifier_len)
+        self.strategy.session_set(name, code_verifier)
+        return code_verifier
```
```suggestion
        self.strategy.session_set(name, code_verifier)
   
        return code_verifier
```

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+        return code_verifier
+
+    def get_code_verifier(self):
+        name = self.name + "_code_verifier"
```
```suggestion
        name = f"{self.name}_code_verifier"
```

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+
+    def get_code_verifier(self):
+        name = self.name + "_code_verifier"
+        code_verifier = self.strategy.session_get(name)
+        return code_verifier
```
```suggestion
        code_verifier = self.strategy.session_get(name)
   
        return code_verifier
```

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+            hashed = hashlib.sha256(code_verifier.encode()).digest()
+            encoded = base64.urlsafe_b64encode(hashed)
+            code_challenge = encoded.decode().replace("=", "")  # remove padding
+            return code_challenge
+        elif method == "plain":
```
```suggestion
            return code_challenge

        if method == "plain":
```

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+            return code_challenge
+        elif method == "plain":
+            return code_verifier
+        else:
+            raise AuthException("Unsupported code challenge method.")
```
```suggestion

        raise AuthException("Unsupported code challenge method.")
```

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+
+    PKCE_DEFAULT_CODE_CHALLENGE_METHOD = "s256"
+    PKCE_DEFAULT_CODE_VERIFIER_LENGTH = 32
+    USE_PKCE = True
```
I don't understand why is this necessary, I mean the obvious answer is because you can disable that behavior on  sub-classes, but if I don't want that behavior on sub-classes I could inherit from  BaseOAuth2 instead of this one

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+                default=self.PKCE_DEFAULT_CODE_CHALLENGE_METHOD,
+            )
+            code_verifier = self.create_code_verifier()
+            code_challenge = self.generate_code_challenge(
+                code_verifier, code_challenge_method
+            )
```
```suggestion
            code_challenge = self.generate_code_challenge(
                code_verifier,
                code_challenge_method,
            )
```

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+                code_verifier, code_challenge_method
+            )
+            params["code_challenge_method"] = code_challenge_method
+            params["code_challenge"] = code_challenge
+        return params
```
```suggestion
            params["code_challenge"] = code_challenge

        return params
```

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+        if method == "s256":
+            hashed = hashlib.sha256(code_verifier.encode()).digest()
+            encoded = base64.urlsafe_b64encode(hashed)
+            code_challenge = encoded.decode().replace("=", "")  # remove padding
+            return code_challenge
```
```suggestion
            return encoded.decode().replace("=", "")  # remove padding
```

andrey-canon commented on eox_core/social_tpa_backends.py:
```diff
@@ -199,3 +201,83 @@ def get_user_id(self, *args, **kwargs):
+    def create_code_verifier(self):
+        name = self.name + "_code_verifier"
+        code_verifier_len = self.setting(
+            "PKCE_CODE_VERIFIER_LENGTH", default=self.PKCE_DEFAULT_CODE_VERIFIER_LENGTH
```
```suggestion
            "PKCE_CODE_VERIFIER_LENGTH",
            default=self.PKCE_DEFAULT_CODE_VERIFIER_LENGTH,
```

andrey-canon reviewed (commented):
LGTM,  I just left some comments related with the code's style that I would have implemented , however I know this is a copy of the social-core pr  so feel free to ignore these I just wanted to point them out

johanv26 commented:
@andrey-canon I updated to the version is approved in social_core. But taking advantage of that PR is not merged yet, I added some of your review to that PR.
https://github.com/python-social-auth/social-core/pull/856#pullrequestreview-1742515055
Maybe they pay attention and would be included in the social_core version.
